stdlib/public/core/Random.swift

```diff
@@ -0,0 +1,167 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+/// A type that can provide uniformly distributed random data.
+///
+/// When you call methods that use random data, such as creating new random
+/// values or shuffling a collection, you can pass a `RandomNumberGenerator`
+/// type to be used as the source for randomness. When you don't pass a
+/// generator, the default `Random` type is used.
+///
+/// When providing new APIs that use randomness, provide a version that accepts
+/// a generator conforming to the `RandomNumberGenerator` protocol as well as a
+/// version that uses the default generator. For example, this `Weekday`
+/// enumeration provides static methods that return a random day of the week:
+///
+///     enum Weekday : CaseIterable {
+///         case sunday, monday, tuesday, wednesday, thursday, friday, saturday
+///
+///         static func randomWeekday<G: RandomNumberGenerator>(using generator: inout G) -> Weekday {
+///             return Weekday.allCases.randomElement(using: &generator)!
+///         }
+///
+///         static func randomWeekday() -> Weekday {
+///             return Weekday.randomWeekday(using: &Random.default)
+///         }
+///     }
+///
+/// Conforming to the RandomNumberGenerator Protocol
+/// ================================================
+///
+/// A custom `RandomNumberGenerator` type can have different characteristics
+/// than the default `Random` type. For example, a seedable generator can be
+/// used to generate the same sequence of random values for testing purposes.
+///
+/// To make a custom type conform to the `RandomNumberGenerator` protocol,
+/// implement the required `next()` method. Each call to `next()` must produce
+/// a uniform and independent random value.
+///
+/// Types that conform to `RandomNumberGenerator` should specifically document
+/// the thread safety and quality of the generator.
+public protocol RandomNumberGenerator {
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: An unsigned 64-bit random value.
+  mutating func next() -> UInt64
+}
+
+extension RandomNumberGenerator {
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: A random value of `T`. Bits are randomly distributed so that
+  ///   every value of `T` is equally likely to be returned.
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>() -> T {
+    if T.bitWidth <= UInt64.bitWidth {
+      return T(truncatingIfNeeded: self.next())
+    }
+
```

Comment:
I wonder if it would make sense to split off the rest of this function into a non-`@inlinable` `@inline(never)` internal method, and `_fastPath` the branch above. This could make it more likely that the 99.99% use case and fast-path is more reliably inlined and/or specialized. CC @eeckstein for thoughts.

Also, how is the rest of this code tested?

Reply:
We have the `DoubleWidth` type here, we could test a 128-bit integer type with that: https://github.com/apple/swift/blob/master/test/Prototypes/DoubleWidth.swift.gyb

If we want to test anything that isn't a power of two wide, that'll be an additional thing to develop.